src/waveform/widgets/allshader/waveformwidget.h

```diff
@@ -42,12 +48,31 @@ class allshader::WaveformWidget final : public ::WGLWidget,
     void wheelEvent(QWheelEvent* event) override;
     void leaveEvent(QEvent* event) override;
 
-    allshader::WaveformRendererSignalBase* addWaveformSignalRenderer(
+    template<class T_Renderer, typename... Args>
+    inline std::unique_ptr<T_Renderer> addRendererNode(Args&&... args) {
+        return std::unique_ptr<T_Renderer>(addRenderer<T_Renderer>(std::forward<Args>(args)...));
+    }
+
+    template<class T_Renderer, typename... Args>
+    inline std::unique_ptr<T_Renderer> addWaveformSignalRendererNode(Args&&... args) {
+        auto pRenderer = addRenderer<T_Renderer>(std::forward<Args>(args)...);
+        if (!m_pWaveformRendererSignal) {
+            m_pWaveformRendererSignal = pRenderer;
+        }
```

Comment:
can you explain why this is only added if with this condition?

Reply:
This template function is called twice, first for the normal signal renderer, and later for the slip signal renderer.

But this is convoluted and not clear. I will try to improve this.

Reply:
this was a bit tricky, but I think I found a relatively elegant solution.